design/20210209.certificates.k8s.io-adoption.md

```diff
@@ -0,0 +1,319 @@
+operation scenarios during installation or upgrades. For example, the
+`CertificateRequest` resource needs to be installed and the webhook ready for
+consumers/clients of this API to make requests.
+
+In some setups, the cert-manager installation is handled by two separate teams:
+
+- The cert-manager CRDs and webhook are installed and managed by the
+  operations team, which allows the operations team to enforce cluster
+  policies;
+- The cert-manager controller, cainjector and acmesolver are installed and
+  managed by the developer team.
+
+By supporting the built-in `CertificateSigningRequest` resource, we decrease the
+coupling between the operations team and the developer team.  Since the
+`CertificateSigningRequest` resource is built-in, it does not need any
+`CustomResourceDefinition` installation, and the webhook for this resource is
+integrated to the API server.
+
+
+### CertificateSigningRequest Migration Path
+
+Support for the `CertificateSigningRequest` resource in cert-manager is a
+stepping stone to migrating to this resource instead of the `CertificateRequest`
+resource, if the project does indeed want to migrate.
+
+
+### Goals
+
+* Add `CertificateSigningRequest` signer controllers to cert-manager for all
+    `Issuer` types
+* Pave a way for replacement of the `CertificateRequest` resource for `CertificateSigningRequest`
+
+### Non-Goals
+
+* Remove the `CertificateRequest` resource before `CertificateSigningRequest` is
+    fully supported and a migration path is in place
+
+---
+
+## Reasons for Approach
+
+Below are a summary of properties of the `CertificateSigningRequest` which
+cause issues or caveats which need to be addressed when adding support for the
+resource.
+
+### Read all access to CertificateSigningRequests
+
+`CertificateSigningRequests` are cluster scoped resources, whereas
+`CertificateRequests` are not. This means that in order for users to have the
+same debugging and usage capabilities of the `CertificateRequest`, users will
+have the ability to list all requests in the cluster, rather than being scoped to
+specific namespaces. This is a regression in security.
+
+A solution to this problem could be given to users by providing examples or
+guides to external tooling (e.g. [Open Policy Agent
+(OPA)](https://github.com/open-policy-agent/opa), whereby cluster administrators
+can limit users access to certain `CertificateSigningRequests` based on label
+selectors.
+
+
+### Namespaced issuers referenced by CertificateSigningRequest
+
+`CertificateSigningRequests` are cluster scoped resources, whereas
+`CertificateRequests` are not. cert-manager has the concept of namespaced
+issuers that may only be referenced by `CertificateRequests` that reside in
+the same namespace. Since `CertificateSigningRequests` can reference any
+namespaced `Issuer`, there must be some mechanism to prevent privilege
+escalation (Alice references an issuer in Bob's namespace).
```

Comment:
The built in admission only supports a simple wildcard, so you would have to encode the namespace into the domain part itself.  That does seem doable though:

`ns1.issuer.cert-manager.io/issuser1` for a single issuer
`ns1.issuer.cert-manager.io/*` for all issuers in the namespace

Reply:
Ignore me, I misremembered how this worked in Kube CSR.  Not enough sleep.